File: Mobile/lib/features/pos/models/sale.dart
class SaleItemSnapshot {
  final String productId;
  final String productNameSnapshot;
  final double quantity;
  final double unitPrice;
  final double vatRate;
  final double vatAmount;
  final double lineTotal;

  SaleItemSnapshot({
    required this.productId,
    required this.productNameSnapshot,
    required this.quantity,
    required this.unitPrice,
    required this.vatRate,
    required this.vatAmount,
    required this.lineTotal,
  });

  factory SaleItemSnapshot.fromJson(Map<String, dynamic> json) => SaleItemSnapshot(
        productId: json['productId'],
        productNameSnapshot: json['productNameSnapshot'] ?? '',
        quantity: (json['quantity'] ?? 0).toDouble(),
        unitPrice: (json['unitPrice'] ?? 0).toDouble(),
        vatRate: (json['vatRate'] ?? 0).toDouble(),
        vatAmount: (json['vatAmount'] ?? 0).toDouble(),
        lineTotal: (json['lineTotal'] ?? 0).toDouble(),
      );
}

class Sale {
  final String id;
  final String invoiceNumber;
  final String? customerId;
  final String? customerName;
  final DateTime saleDate;
  final double subTotal;
  final double discountAmount;
  final double vatAmount;
  final double total;
  final double paidAmount;
  final double changeAmount;
  final int status;
  final String? eInvoiceUuid;
  final int? eInvoiceStatus;
  final List<SaleItemSnapshot> items;

  Sale({
    required this.id,
    required this.invoiceNumber,
    this.customerId,
    this.customerName,
    required this.saleDate,
    required this.subTotal,
    required this.discountAmount,
    required this.vatAmount,
    required this.total,
    required this.paidAmount,
    required this.changeAmount,
    required this.status,
    this.eInvoiceUuid,
    this.eInvoiceStatus,
    required this.items,
  });

  factory Sale.fromJson(Map<String, dynamic> json) => Sale(
        id: json['id'],
        invoiceNumber: json['invoiceNumber'] ?? '',
        customerId: json['customerId'],
        customerName: json['customerName'],
        saleDate: DateTime.parse(json['saleDate']),
        subTotal: (json['subTotal'] ?? 0).toDouble(),
        discountAmount: (json['discountAmount'] ?? 0).toDouble(),
        vatAmount: (json['vatAmount'] ?? 0).toDouble(),
        total: (json['total'] ?? 0).toDouble(),
        paidAmount: (json['paidAmount'] ?? 0).toDouble(),
        changeAmount: (json['changeAmount'] ?? 0).toDouble(),
        status: json['status'] ?? 0,
        eInvoiceUuid: json['eInvoiceUuid'],
        eInvoiceStatus: json['eInvoiceStatus'],
        items: ((json['items'] ?? []) as List)
            .map((e) => SaleItemSnapshot.fromJson(e))
            .toList(),
      );
}

File: Domain/Models/POS/Sale.cs
using System.ComponentModel.DataAnnotations;
using Domain.Enums;
using Domain.Models.Inventory;

namespace Domain.Models.POS
{
    public class Sale
    {
        [Key]
        public Guid Id { get; set; } = Guid.NewGuid();

        [Required, StringLength(50)]
        public string InvoiceNumber { get; set; } = string.Empty;

        public Guid? CustomerId { get; set; }
        public Customer? Customer { get; set; }

        public Guid WarehouseId { get; set; }
        public Warehouse? Warehouse { get; set; }

        public Guid CashSessionId { get; set; }
        public CashSession? CashSession { get; set; }

        public Guid CashierUserId { get; set; }

        public DateTime SaleDate { get; set; } = DateTime.UtcNow;

        public decimal SubTotal { get; set; }
        public decimal DiscountAmount { get; set; }
        public decimal DiscountPercent { get; set; }

        public decimal VatAmount { get; set; }
        public decimal Total { get; set; }
        public decimal PaidAmount { get; set; }
        public decimal ChangeAmount { get; set; }

        public SaleStatus Status { get; set; } = SaleStatus.Completed;

        // Egyptian e-invoicing reference
        [StringLength(100)]
        public string? EInvoiceUuid { get; set; }
        public EInvoiceStatus? EInvoiceStatus { get; set; }
        [StringLength(100)]
        public string? EInvoiceLongId { get; set; }

        [StringLength(500)]
        public string? Notes { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public ICollection<SaleItem>? Items { get; set; }
        public ICollection<SalePayment>? Payments { get; set; }
    }
}
